enarx-keep/tests/integration_tests.rs

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
+use std::time::Duration;
+use wait_timeout::ChildExt;
+
+const CRATE: &str = env!("CARGO_MANIFEST_DIR");
+const OUT_DIR: &str = env!("OUT_DIR");
+const TEST_DIR: &str = "test-bin";
+const TIMEOUT_SECS: u64 = 5;
+
+fn path(input: &str) -> PathBuf {
+    Path::new(CRATE).join(OUT_DIR).join(TEST_DIR).join(input)
+}
+
+/// Initializes a test with filtered environment.
+fn test_init() {
+    let mut filtered_env: HashMap<String, String> = std::env::vars()
+        .filter(|&(ref k, _)| {
+            k == "TERM"
+                || k == "TZ"
+                || k == "LANG"
+                || k == "PATH"
+                || k == "RUSTUP_HOME"
+                || k == "RUSTFLAGS"
+        })
+        .collect();
+
+    filtered_env.insert("ENARX_INTEGRATION_TESTS".into(), "1".into());
+
+    let status = Command::new("cargo")
+        .current_dir(CRATE)
+        .env_clear()
+        .envs(filtered_env)
+        .arg("+nightly")
+        .arg("build")
+        .stdout(Stdio::null())
+        .stderr(Stdio::null())
+        .status()
+        .expect("failed to prepare");
+
+    if !status.success() {
+        eprintln!("failed to prepare");
+        std::process::exit(1);
+    }
+}
+
+/// Returns true if running the binary exits with 0. Used in cases
+/// where there is no output or output does not matter.
+fn run_succeeds(bin: PathBuf) -> bool {
+    let mut filtered_env: HashMap<String, String> = std::env::vars()
+        .filter(|&(ref k, _)| {
+            k == "TERM"
+                || k == "TZ"
+                || k == "LANG"
+                || k == "PATH"
+                || k == "RUSTUP_HOME"
+                || k == "RUSTFLAGS"
+        })
+        .collect();
+
+    filtered_env.insert("ENARX_INTEGRATION_TESTS".into(), "1".into());
+
+    let mut cmd = Command::new("cargo")
+        .current_dir(CRATE)
+        .arg("+nightly")
+        .env_clear()
+        .envs(filtered_env)
+        .arg("run")
+        .arg("-q")
+        .arg("exec")
+        .arg(bin)
+        .spawn()
+        .expect("failed to run test bin");
```

Comment:
This is why you're having so much problem with extra output and having to work around it with a bunch of hacks. When `cargo test` is run, the binary is already built before the tests are run. You just need to execute it directly rather than using `cargo +nightly run`.

I created a [working example of this strategy](https://github.com/npmccallum/demo/). We execute the binary [like this](https://github.com/npmccallum/demo/blob/master/tests/test.rs#L7-L8). This gives no extra output we have to work around.

In short, if you run the binary directly and don't use `cargo run` you can probably remove half the code in this commit which works around the problems of using `cargo run`.

Reply:
@npmccallum Thank you!!!